opa/pkg/main.go

```diff
@@ -0,0 +1,324 @@
+		check, err := dynamicInterface.Resource(checkGvr).Namespace(checkInstanceObject.Namespace).Get(ctx, checkInstanceObject.Spec.CustomCheckName, metav1.GetOptions{})
+		if err != nil {
+			actionItems = append(actionItems, ActionItem{
+				Title:       checkInstanceObject.Name,
+				Description: "An error occured retrieving the Custom Check for this instance",
+				Remediation: "Make sure that the custom check exists and it is in the same namespace as this instance.",
+				Severity:    0.4,
+				Category:    "Reliability",
+			})
+			logrus.Warnf("Error while retrieving check %+v", err)
+			continue
+		}
+		var checkObject customCheck
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(check.Object, &checkObject)
+		if err != nil {
+			return nil, err
+		}
+
+		newItems, err := processCheck(ctx, checkObject, checkInstanceObject, restMapper, dynamicInterface)
+		if err != nil {
+			return nil, err
+		}
+		actionItems = append(actionItems, newItems...)
+	}
+	return actionItems, nil
+}
+
+func getKubernetesDataFunction(ctx context.Context, check customCheck, restMapper meta.RESTMapper, dynamicInterface dynamic.Interface) func(rego.BuiltinContext, *ast.Term) (*ast.Term, error) {
+	return func(_ rego.BuiltinContext, a *ast.Term) (*ast.Term, error) {
+
+		str, ok := a.Value.(ast.String)
+		if !ok {
+			return nil, nil
+		}
+		strValue := str.String()
+		if len(strValue) > 0 && strValue[0] == '"' {
+			strValue = strValue[1:]
+		}
+		if len(strValue) > 0 && strValue[len(strValue)-1] == '"' {
+			strValue = strValue[:len(strValue)-1]
+		}
+		var apiGroup string
+		for _, target := range check.Spec.AdditionalKubernetesData {
+			fmt.Printf("Checking if %s matches %s", strValue, target.Kinds[0])
+			if strValue == target.Kinds[0] {
+				fmt.Printf("It does!")
+				apiGroup = target.APIGroups[0]
+				break
+			}
+		}
+		mapping, err := restMapper.RESTMapping(schema.GroupKind{Group: apiGroup, Kind: strValue})
+		if err != nil {
+			return nil, err
+		}
+		gvr := mapping.Resource
+		list, err := dynamicInterface.Resource(gvr).Namespace("").List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, err
+		}
+		items := make([]interface{}, 0)
+		for _, item := range list.Items {
+			items = append(items, item.Object)
+		}
+		itemValue, err := ast.InterfaceToValue(items)
+		if err != nil {
+			return nil, err
+		}
+
+		return ast.NewTerm(itemValue), nil
+	}
+
+}
+func processCheck(ctx context.Context, check customCheck, checkInstance customCheckInstance, restMapper meta.RESTMapper, dynamicInterface dynamic.Interface) ([]ActionItem, error) {
+	actionItems := make([]ActionItem, 0)
+
+	getKubernetesData := getKubernetesDataFunction(ctx, check, restMapper, dynamicInterface)
+	for _, target := range checkInstance.Spec.Targets {
+		for _, apiGroup := range target.APIGroups {
+			for _, kind := range target.Kinds {
+				logrus.Infof("Starting to process %s %s\n\n", apiGroup, kind)
```

Comment:
this is still a little hard to read on GH - any chance we can break out the inner loop into its own function?

Reply:
I did a bunch of refactoring, should remove a ton of indentation.

Reply:
Thanks!